tactics/class_tactics.ml

```diff
@@ -1088,14 +1094,24 @@ module Search = struct
       let derivs = path_derivate info.search_cut name in
       let pr_error ie =
         if !typeclasses_debug > 1 then
-          let msg =
-            pr_depth (!idx :: info.search_depth) ++ str": " ++
+          let idx = if fst ie == NoApplicableEx then pred !idx else !idx in
+          let header =
+            pr_depth (idx :: info.search_depth) ++ str": " ++
               Lazy.force pp ++
               (if !foundone != true then
                  str" on" ++ spc () ++ pr_ev s (Proofview.Goal.goal (Proofview.Goal.assume gl))
                else mt ())
           in
-          Feedback.msg_debug (msg ++ str " failed with " ++ CErrors.iprint ie)
+          let msg =
+            match fst ie with
+            | Pretype_errors.PretypeError (env, evd, Pretype_errors.CannotUnify (x,y,_)) ->
+               str"Cannot unify " ++ print_constr_env env evd x ++ str" and " ++
+                                           print_constr_env env evd y
+            | ReachedLimitEx -> str "Proof-search reached its limit."
+            | NoApplicableEx -> str "Proof-search failed."
+            | e -> CErrors.iprint ie
+          in
+          Feedback.msg_debug (header ++ str " failed with: " ++ msg)
```

Comment:
Why not use `Himsg.explain_pretype_error` for the first case? Although I guess you don't really many details, since this output is only some debug information.

Apart from that point, I am not a reviewer, but it looks fine to me.